contracts/drafts/ERC777/ERC777.sol

```diff
@@ -0,0 +1,402 @@
+    }
+
+    /**
+    * @dev Gets the token's granularity,
+    * i.e. the smallest number of tokens (in the basic unit)
+    * which may be minted, sent or burned at any time
+    * @return uint256 granularity
+     */
+    function granularity() public view returns (uint256) {
+        return _granularity;
+    }
+
+    /**
+    * @dev Get the list of default operators as defined by the token contract.
+    * @return address[] default operators
+    */
+    function defaultOperators() public view returns (address[] memory) {
+        return _defaultOperatorsArray;
+    }
+
+    /**
+    * @dev Authorize an operator for the sender
+    * @param operator address to be authorized as operator
+     */
+    function authorizeOperator(address operator) public {
+        require(msg.sender != operator);
+        if (_defaultOperators[operator]) {
+            _reAuthorizeDefaultOperator(operator);
+        } else {
+            _authorizeOperator(operator);
+        }
+    }
+
+    /**
+     * @dev Revoke operator rights from one of the default operators
+     * @param operator address to revoke operator rights from
+     */
+    function revokeOperator(address operator) public {
+        require(operator != msg.sender);
+        if (_defaultOperators[operator]) {
+            _revokeDefaultOperator(operator);
+        } else {
+            _revokeOperator(operator);
+        }
+    }
+
+    /**
+    * @dev Indicate whether an address
+    * is an operator of the tokenHolder address
+    * @param operator address which may be an operator of tokenHolder
+    * @param tokenHolder address of a token holder which may have the operator
+    * address as an operator.
+     */
+    function isOperatorFor(
+        address operator,
+        address tokenHolder
+    ) public view returns (bool) {
+        return operator == tokenHolder ||
+            _defaultOperators[operator] && !_revokedDefaultOperators[tokenHolder][operator] ||
+            _operators[tokenHolder][operator];
+    }
+
+    /**
+     * @dev Mint tokens. Does not check authorization of operator
+     * @dev the caller may ckeck that operator is authorized before calling
+     * @param operator address operator requesting the operation
+     * @param to address token recipient address
+     * @param amount uint256 amount of tokens to mint
+     * @param userData bytes extra information defined by the token recipient (if any)
+     * @param operatorData bytes extra information provided by the operator (if any)
+     */
+    function _mint(
+        address operator,
+        address to,
+        uint256 amount,
+        bytes memory userData,
+        bytes memory operatorData
+    )
+    internal
+    {
+        require(to != address(0));
+        require((amount % _granularity) == 0);
+
+        // revert if 'to' is a contract not implementing tokensReceived()
+        require(_callTokensReceived(operator, address(0), to, amount, userData, operatorData));
```

Comment:
The ordering of this is probably wrong. Although the EIP is not explicit about it for minting, it's very clear in `send` that the `tokensReceived` hook has to be called after updating the balance:

>The token contract MUST call the `tokensReceived` hook after updating the state.

Reply:
It actually is explicit:

>>>The following rules apply when calling the tokensReceived hook:
    The tokensReceived hook MUST be called every time the balance is incremented.
    The tokensReceived hook MUST be called after the state is updated—i.e. after the balance is incremented.